### DISCUSSION TITLE
Is [DataMember(IsRequired=true)] supported?

### DISCUSSION THREAD

**Discussion by mshann03** [CoreWCF/CoreWCF#677] [Q&A] [answered]
Hi, 

I can't seem to get the serialization of incoming coreWCF requests to fail if a property is marked as required and I don't include that element in the payload.

For example, if we look at https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Http/tests/HttpClientTests.cs

If we change the `EchoMessage` contract to include this attribute, like:
```
        [DataContract]
        public class EchoMessage
        {
            [DataMember(IsRequired = true)]
            public string Text { get; set; }
        }
```

And then modify line 42 to this:
```
const string requestBody = @"<s:Envelope xmlns:s=""http://schemas.xmlsoap.org/soap/envelope/"" xmlns:tem=""http://tempuri.org/"">
   <s:Header/>
   <s:Body>
      <tem:Echo />
   </s:Body>
</s:Envelope>";
````

The echo method gets called passing in null as the text. I would have thought the serialization of the payload would fail.

Is this as expected or a bug?

**Comment by mconnew** [comment, marked as answer]
That test isn't using the EchoMessage class. This is the definition of the service interface:
```c#
        [ServiceContract]
        public interface IEchoService
        {
            [OperationContract]
            string Echo(string text);
        }
```
As you can see, it's passing a string called text. If you want to test your scenario, you need to modify the interface and implementation to use EchoMessage:
### IEchoService
```c#
        [ServiceContract]
        public interface IEchoService
        {
            [OperationContract]
            string Echo(EchoMessage echo);
        }
```
### EchoService
```c#
        public class EchoService : IEchoService
        {
            public string Echo(EchoMessage echo)
            {
                Console.WriteLine($"Received {echo.Text} from client!");
                return echo.Text;
            }
        }
```
You then need to modify the request string that's sent to look like this:
```c#
            const string requestBody = @"<s:Envelope xmlns:s=""http://schemas.xmlsoap.org/soap/envelope/"" xmlns:tem=""http://tempuri.org/"">
   <s:Header/>
   <s:Body>
      <tem:Echo>
        <tem:echo xmlns:i=""http://www.w3.org/2001/XMLSchema-instance"" xmlns=""http://schemas.datacontract.org/2004/07/CoreWCF.Http.Tests"">
          <Text>A</Text>
        </tem:echo>
      </tem:Echo>
   </s:Body>
</s:Envelope>";
```
Running the test still passes as the xml has the Text element with an A. If you don't have the `IsRequired = true` property setter on the `DataMember` attribute, removing the `<Text>A</Text>` node will result in the service receiving a `null` value in the `Text` property and returning `null`. If you add the `IsRequired = true` property setter and remove the `Text` node from the xml, you get the following response message:
```xml
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
	<s:Body>
		<s:Fault>
			<faultcode xmlns:a="http://schemas.microsoft.com/net/2005/12/windowscommunicationfoundation/dispatcher">a:DeserializationFailed</faultcode>
			<faultstring xml:lang="en-US">The formatter threw an exception while trying to deserialize the message: There was an error while trying to deserialize parameter http://tempuri.org/:echo. The InnerException message was 'Error in line 7 position 20. 'EndElement' 'echo' from namespace 'http://tempuri.org/' is not expected. Expecting element 'Text'.'.  Please see InnerException for more details.</faultstring>
		</s:Fault>
	</s:Body>
</s:Envelope>
```

I believe this is the same behavior from .NET Framework. Are you expecting something different?
```

**Reply by mshann03** [reply]
Apologies @mconnew for sending through an incomplete example. I did mean to pass through the EchoMessage in the operation - I've been working on my own example which did just that, and then when I posted the question retrofitted to your example I didn't implement it right. 

Long story short, you are correct and I found out the issue on my side was we had decorated the contract with `[XmlSerializerFormat]` which turns out to ignore the DCS specific attributes. 

Sorry for not realising this sooner. Thanks for the response.
